lib/burnside/intrinsics.cc

```diff
@@ -448,4 +460,61 @@ mlir::Value *IntrinsicLibrary::Implementation::genConjg(
   return cplxHandler.insert<ComplexHandler::Part::Imag>(cplx, negImag);
 }
 
+static mlir::FuncOp getMergeFunc(mlir::Type type, mlir::ModuleOp module) {
+  llvm::SmallVector<mlir::Type, 3> argTypes{
+      type, type, fir::LogicalType::get(module.getContext(), 4)};
+  auto mergeType{
+      mlir::FunctionType::get(argTypes, {type}, module.getContext())};
+  auto mergeName{getIntrinsicWrapperName("merge", mergeType)};
+  return createFunction(module, mergeName, mergeType);
+}
+
+template<Extremum extremum>
+static mlir::Value *createCompare(mlir::Location loc, mlir::OpBuilder &builder,
+    mlir::Value *left, mlir::Value *right) {
+  static constexpr auto integerPredicate{extremum == Extremum::Max
+          ? mlir::CmpIPredicate::SGT
+          : mlir::CmpIPredicate::SLT};
+  static constexpr auto realPredicate{extremum == Extremum::Max
+          ? fir::CmpFPredicate::UGT
+          : fir::CmpFPredicate::ULT};
```

Comment:
Is "unordered or greater-than" the right test here?  It would mean MAX returns NAN if at least one argument is a NAN.

Reply:
>  I think this behavior would be obtained for f18 using olt and ogt in the code just above.

Can we prove this? Since 3 out of 6 compilers suggest that behavior, we can claim herd immunity. :)

Reply:
Yes, given the implementation is based on merge, when the comparison is true, the first argument is returned. So, using `ult` and `ugt` (resps `ogt` and `olt`), the first (resp. second) argument is always returned if one of the argument is a nan. So we can implement all the behaviors of other compilers as:

pgfortran:
```
MIN(a, b) = MERGE(a,b, a ult b)
MAX(a, b) = MERGE(a,b, a ugt b)
```
gfortran (this one is more comlpex):
```
MIN= MERGE(a, b, a oeq a & a ult b)
MAX= MERGE(a, b, a oeq a & a ugt b)
```
ifort, nagfor, pgfortran -Mnollvm:
```
MIN(a, b) = MERGE(a,b, a olt b)
MAX(a, b) = MERGE(a,b, a ogt b)
```
xlf:
```
MIN(a, b) = MERGE(a,b, a ult b)
MAX(a, b) = MERGE(a,b, a ogt b)
```

I am fine with siding with the majority here, we may want to document this kind of choice somewhere.
I see one cons for this choice being that with pgfortran (with llvm), we could skip evaluation of all the max/min arguments after getting a NaN argument (but that would require inserting short-circuiting logic).

Reply:
Sounds good to me. I'd document this as a comment in the code so the decision and rationale used to reach it doesn't get lost.